Python/ast_opt.c

```diff
@@ -624,20 +624,87 @@ static int
 fold_compare(expr_ty node, PyArena *arena, _PyASTOptimizeState *state)
 {
     asdl_int_seq *ops;
-    asdl_expr_seq *args;
+    asdl_expr_seq *args, *elts;
+    expr_ty arg;
     Py_ssize_t i;
+    Py_ssize_t elts_len;
 
     ops = node->v.Compare.ops;
     args = node->v.Compare.comparators;
     /* Change literal list or set in 'in' or 'not in' into
        tuple or frozenset respectively. */
     i = asdl_seq_LEN(ops) - 1;
     int op = asdl_seq_GET(ops, i);
+    arg = asdl_seq_GET(args, i);
+    _Bool is_lhs_constant = (node->v.Compare.left->kind) == Constant_kind;
+    //lhs_const = node->v.Compare.left->v.Constant.value;
     if (op == In || op == NotIn) {
-        if (!fold_iter((expr_ty)asdl_seq_GET(args, i), arena, state)) {
-            return 0;
+        if (arg->kind == List_kind) {
+            asdl_expr_seq *list_elts = arg->v.List.elts;
+            if (has_starred(list_elts))
+                return 1;
+            expr_context_ty ctx = arg->v.List.ctx;
+            arg->kind = Tuple_kind;
+            arg->v.Tuple.elts = list_elts;
+            arg->v.Tuple.ctx = ctx;
+            elts = list_elts;
+        }
+        else if (arg->kind == Dict_kind) {
+            elts = arg->v.Dict.keys;
+            arg->kind = Set_kind;
+            arg->v.Set.elts = elts;
+        }
+        else if (arg->kind == Set_kind) {
+            elts = arg->v.Set.elts;
+        }
+        else {
+            return 1;
+        }
+
+        elts_len = asdl_seq_LEN(elts);
+        if (!elts_len) {
+            return make_const(node, op == In ? Py_False : Py_True, arena);
+        }
+
+        PyObject *newval = PyTuple_New(elts_len);
+        if (newval == NULL) {
+            if (PyErr_ExceptionMatches(PyExc_KeyboardInterrupt)) {
```

Comment:
There is no need to match against `PyExc_KeyboardInterrupt` here. Just treat this like a regular error and bubble up